privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,190 @@
+        :type db_token:  orm object
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"tan")
+        self.hKeyRequired = False
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the token type shortname
+
+        :return: 'paper'
+        :rtype: string
+        """
+        return "tan"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        Return the prefix, that is used as a prefix for the serial numbers.
+        :return: PITN
+        """
+        return "PITN"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        returns a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or scalar
+        """
+        res = {'type': 'tan',
+               'title': 'TAN Token',
+               'description': 'TAN: TANs printed on a sheet '
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
+                   }
+               }
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    def update(self, param, reset_failcount=True):
+        tan_dict = {}
+        if "tans" in param:
+            # init tokens with tans
+            tans = param.get("tans").split()
+            tan_dict = {k: v for k, v in enumerate(tans)}
+        else:
+            # Init token without tans, so we create tans
+            param["papertoken_count"] = param.get("tantoken_count") or DEFAULT_COUNT
+            PaperTokenClass.update(self, param, reset_failcount=reset_failcount)
+            # After this creation, the init_details contain the complete list of the TANs
+            tan_dict = self.init_details.get("otps", {})
+        for tankey, tanvalue in tan_dict.iteritems():
+            # Get a 4 byte salt from the crypto module
+            salt = geturandom(SALT_LENGTH, hex=True)
+            # Now we add all TANs to the tokeninfo of this token.
+            hashed_tan = binascii.hexlify(hash(tanvalue, salt))
+            self.add_tokeninfo("tan.tan{0!s}".format(tankey),
+                               "{0}:{1}".format(binascii.hexlify(salt),
```

Comment:
Hmm, shouldn't the salt already be a hexstring because we passed ``hex=True``?

Reply:
You are right. Done.